rllib/utils/explorations/epsilon_greedy.py

```diff
@@ -0,0 +1,177 @@
+            self.last_time_step + 1
+
+        if explore:
+            # Get the current epsilon.
+            epsilon = self.epsilon_schedule(self.last_time_step)
+
+            batch_size = q_values.size()[0]
+            # Mask out actions, whose Q-values are -inf, so that we don't
+            # even consider them for exploration.
+            random_valid_action_logits = torch.where(
+                q_values == float("-inf"),
+                torch.ones_like(q_values) * float("-inf"),
+                torch.ones_like(q_values))
+
+            random_actions = torch.squeeze(
+                torch.multinomial(random_valid_action_logits, 1), axis=1)
+
+            return torch.where(
+                torch.empty((batch_size,)).uniform_() < epsilon,
+                random_actions, action)
+
+        # Return the action.
+        else:
+            return action
+
+    def _get_tf_exploration_action_op(self, action, explore, time_step,
+                                      q_values):
+        """
+        Tf helper method to produce the tf op for an epsilon exploration
+            action.
+
+        Args:
+            action (tf.Tensor): The already sampled action (non-exploratory
+                case) as tf op.
+
+        Returns:
+            tf.Tensor: The tf exploration-action op.
+        """
+        epsilon = tf.convert_to_tensor(
+            self.epsilon_schedule(time_step if time_step is not None else
+                                  self.last_time_step))
+
+        batch_size = tf.shape(action)[0]
+
+        # Maske out actions with q-value=-inf so that we don't
+        # even consider them for exploration.
+        random_valid_action_logits = tf.where(
+            tf.equal(q_values, tf.float32.min),
+            tf.ones_like(q_values) * tf.float32.min, tf.ones_like(q_values))
+        random_actions = tf.squeeze(
+            tf.multinomial(random_valid_action_logits, 1), axis=1)
+
+        chose_random = tf.random_uniform(
+            tf.stack([batch_size]),
+            minval=0, maxval=1, dtype=epsilon.dtype) \
+            < epsilon
+
+        exploration_action = tf.cond(
+            pred=explore,
+            true_fn=lambda: tf.where(chose_random, random_actions, action),
+            false_fn=lambda: action,
+        )
+        # Increment `last_time_step` by 1 (or set to `time_step`).
+        assign_op = \
+            tf.assign_add(self.last_time_step, 1) if time_step is None else \
+            tf.assign(self.last_time_step, time_step)
+        with tf.control_dependencies([assign_op]):
+            return exploration_action
+
+    @override(Exploration)
+    def get_info(self):
+        """
+        Returns:
+            Union[float,tf.Tensor[float]]: The current epsilon value.
+        """
+        return [self.epsilon_schedule(self.last_time_step)]
+
+    @override(Exploration)
+    def get_state(self):
+        return [self.last_time_step]
```

Comment:
We don't need this right? Isn't it set automatically already via global vars in the policy (and passed in via get_action as the current time step)?